storage/aptosdb/src/state_store/mod.rs

```diff
@@ -1169,18 +1172,40 @@ impl StateValueWriter<StateKey, StateValue> for StateStore {
             self.state_kv_db.enabled_sharding(),
         )?;
         self.state_kv_db
-            .commit(version, batch, sharded_schema_batch)
+            .commit(version, batch, sharded_schema_batch)?;
+        if self.internal_indexer_db.is_some() {
+            let keys = node_batch.iter().map(|(key, _)| key.0.clone()).collect();
+            self.internal_indexer_db
+                .as_ref()
+                .unwrap()
+                .write_keys_to_indexer_db(&keys, version, progress)?;
+        }
+        Ok(())
     }
 
     fn write_usage(&self, version: Version, usage: StateStorageUsage) -> Result<()> {
         self.ledger_db.metadata_db().put_usage(version, usage)
     }
 
     fn get_progress(&self, version: Version) -> Result<Option<StateSnapshotProgress>> {
-        Ok(self
+        let main_db_progress = self
             .state_kv_db
             .metadata_db()
             .get::<DbMetadataSchema>(&DbMetadataKey::StateSnapshotRestoreProgress(version))?
-            .map(|v| v.expect_state_snapshot_progress()))
+            .map(|v| v.expect_state_snapshot_progress());
+
+        if self.internal_indexer_db.is_some() {
+            let progress_opt = self
+                .internal_indexer_db
+                .as_ref()
+                .unwrap()
+                .get_restore_progress(version)?;
+            ensure!(
```

Comment:
we probably cannot ensure here because they are not written in a atomic transaction?

Reply:
changed to panic.

Reply:
updated to check if main_db_progress > internal_indexer_progress based on offline discussion